modules/hal_nordic/CMakeLists.txt

```diff
@@ -1,8 +1,117 @@
+    )
+endif ()
+
+target_compile_definitions(nrf-802154-platform
+  PUBLIC
+    # Radio driver shim layer uses raw api
+    NRF_802154_USE_RAW_API=1
+
+    # Number of slots containing short addresses of nodes for which
+    # pending data is stored.
+    NRF_802154_PENDING_SHORT_ADDRESSES=${CONFIG_NRF_802154_PENDING_SHORT_ADDRESSES}
+
+    # Number of slots containing extended addresses of nodes for which
+    # pending data is stored.
+    NRF_802154_PENDING_EXTENDED_ADDRESSES=${CONFIG_NRF_802154_PENDING_EXTENDED_ADDRESSES}
+
+    # Number of buffers in receive queue.
+    NRF_802154_RX_BUFFERS=${CONFIG_NRF_802154_RX_BUFFERS}
+
+    # CCA mode options
+    NRF_802154_CCA_CORR_LIMIT_DEFAULT=${CONFIG_NRF_802154_CCA_CORR_LIMIT}
+    NRF_802154_CCA_CORR_THRESHOLD_DEFAULT=${CONFIG_NRF_802154_CCA_CORR_THRESHOLD}
+    NRF_802154_CCA_ED_THRESHOLD_DEFAULT=${CONFIG_NRF_802154_CCA_ED_THRESHOLD}
+
+    # Enable CSMA/CA
+    NRF_802154_CSMA_CA_ENABLED=1
+    NRF_802154_TX_STARTED_NOTIFY_ENABLED=1
+
+    # ACK timeout
+    NRF_802154_ACK_TIMEOUT_ENABLED=1
+)
+
+if (CONFIG_NRF_802154_CCA_MODE_ED)
+  target_compile_definitions(nrf-802154-platform PUBLIC NRF_802154_CCA_MODE_DEFAULT=NRF_RADIO_CCA_MODE_ED)
+elseif (CONFIG_NRF_802154_CCA_MODE_CARRIER)
+  target_compile_definitions(nrf-802154-platform PUBLIC NRF_802154_CCA_MODE_DEFAULT=NRF_RADIO_CCA_MODE_CARRIER)
+elseif (CONFIG_NRF_802154_CCA_MODE_CARRIER_AND_ED)
+  target_compile_definitions(nrf-802154-platform PUBLIC NRF_802154_CCA_MODE_DEFAULT=NRF_RADIO_CCA_MODE_CARRIER_AND_ED)
+elseif (CONFIG_NRF_802154_CCA_MODE_CARRIER_OR_ED)
+  target_compile_definitions(nrf-802154-platform PUBLIC NRF_802154_CCA_MODE_DEFAULT=NRF_RADIO_CCA_MODE_CARRIER_OR_ED)
+endif()
+
+if (NOT CONFIG_IEEE802154_NRF5 AND NOT CONFIG_IEEE802154_NRF5_EXT_IRQ_MGMT)
+  target_compile_definitions(nrf-802154-platform PUBLIC NRF_802154_INTERNAL_RADIO_IRQ_HANDLING=1)
+else()
+  target_compile_definitions(nrf-802154-platform PUBLIC NRF_802154_INTERNAL_RADIO_IRQ_HANDLING=0)
+endif()
+
+set(NRF52_SERIES  ${CONFIG_SOC_SERIES_NRF52X}        CACHE INTERNAL "")
+set(NRF53_SERIES  ${CONFIG_SOC_SERIES_NRF53X}        CACHE INTERNAL "")
+set(SERIALIZATION ${CONFIG_NRF_802154_SERIALIZATION} CACHE INTERNAL "")
+set(SER_HOST      ${CONFIG_NRF_802154_SER_HOST}      CACHE INTERNAL "")
+set(RADIO_DRIVER  ${CONFIG_NRF_802154_RADIO_DRIVER}  CACHE INTERNAL "")
+set(SL_OPENSOURCE ${CONFIG_NRF_802154_SL_OPENSOURCE} CACHE INTERNAL "")
+
+add_subdirectory_ifdef(CONFIG_NRF_802154_SOURCE_HAL_NORDIC ${ZEPHYR_CURRENT_MODULE_DIR}/drivers/nrf_802154 nrf_802154)
+
+target_include_directories(zephyr_interface INTERFACE
+  $<TARGET_PROPERTY:nrf-802154,INTERFACE_INCLUDE_DIRECTORIES>
+)
+target_compile_definitions(zephyr_interface INTERFACE
+  $<TARGET_PROPERTY:nrf-802154,INTERFACE_COMPILE_DEFINITIONS>
+)
+
+if (CONFIG_NRF_802154_RADIO_DRIVER)
+  zephyr_append_cmake_library(nrf-802154-driver)
+  zephyr_append_cmake_library(nrf-802154-sl)
```

Comment:
The symbols contained in `nrf-802154-{driver,sl,serialization}.a` must be visible to other zephyr modules, for example the https://github.com/zephyrproject-rtos/zephyr/blob/master/drivers/ieee802154/ieee802154_nrf5.c
Without appending the cmake library, the libraries will not get `whole-archive`'d and it is not guaranteed that the file mentioned earlier will see the symbols during linking.

Reply:
well, whole archive is not there in order to handle circular dependency (although it also achieve that).
It's main reason is for keeping unused symbol, and allow the linker to decide if those symbols should be kept.
SYS_INIT defined symbols are example of such symbols that would be dropped by linker if not whole-archived.

Could you for a start place a comment that this is purely to ensure symbols can be resolved, and why regular `target_link_library` is not used in this case.

Let's keep for now, and we can see if we want to cleanup on this.
After all, we are mixing pure CMake libs with Zephyr CMake libraies, and should consider if this is the most appropriate way for such situations.

If this was a `libapp` link issue, we have the `APP_LINK_WITH_<lib>` pattern, but that doesn't purely apply to this situation.

Reply:
I think there is one more reason. The 802154 libraries define some weak symbols that must be superseded by proper implementations in the file I mentioned earlier. I will write down that reason in there.

Reply:
great, thanks for the clarification.

Reply:
> The 802154 libraries define some weak symbols that may be optionally superseded by proper implementations in the file I mentioned earlier. 

actually, if the library is considered by the linker, any strong implementations of the weak symbol should be used.
If this doesn't happen, then it's a sign that the linker haven't searched for the symbol, which may happen if the weak symbol is defined in the same file it's being used, and that file has no other link dependencies to where the strong symbol is located.

Could you give some more details ?
What symbols are not correctly linked in.

Reply:
One of the examples of weak symbols is `nrf_802154_received_timestamp` and many others with weak implementations located in `nrf_802154.c`. The object file is archived in `libnrf-802154-driver.a` library. Some of those symbols are referenced from the same translation unit and almost all of them are referenced from other translation units archived in that same static library.
The strong implementations of some of those symbols are located in `ieee802154_nrf5.c` in Zephyr.

From what I observe `ld` first resolves the references between object files within one static library and then only attempts to resolve the unresolved symbols from other sources. Since the weak symbols happen to be defined and used within the same static library, the `ld` doesn't bother with linking strong symbols them unless `whole-archive` is given.